sdk/storage/src/common/crypt.cpp

```diff
@@ -0,0 +1,188 @@
+// SPDX-License-Identifier: MIT
+
+#include "common/crypt.hpp"
+
+#include "common/storage_common.hpp"
+
+#ifdef _WIN32
+#include <Windows.h>
+#include <bcrypt.h>
+#else
+#include <openssl/bio.h>
+#include <openssl/buffer.h>
+#include <openssl/evp.h>
+#include <openssl/hmac.h>
+#endif
+
+#include <stdexcept>
+
+namespace Azure { namespace Storage {
+
+#ifdef _WIN32
+  std::string HMAC_SHA256(const std::string& text, const std::string& key)
+  {
+    struct AlgorithmProviderInstance
+    {
+      BCRYPT_ALG_HANDLE Handle;
+      std::size_t ContextSize;
+      std::size_t HashLength;
+
+      AlgorithmProviderInstance()
+      {
+        NTSTATUS status = BCryptOpenAlgorithmProvider(
+            &Handle, BCRYPT_SHA256_ALGORITHM, NULL, BCRYPT_ALG_HANDLE_HMAC_FLAG);
+        if (!BCRYPT_SUCCESS(status))
+        {
+          throw std::runtime_error("failed to call bcrypt function");
+        }
+        DWORD objectLength = 0;
+        DWORD dataLength = 0;
+        status = BCryptGetProperty(
+            Handle,
+            BCRYPT_OBJECT_LENGTH,
+            (PBYTE)&objectLength,
+            sizeof(objectLength),
+            &dataLength,
+            0);
+        if (!BCRYPT_SUCCESS(status))
+        {
+          throw std::runtime_error("failed to call bcrypt function");
+        }
+        ContextSize = objectLength;
+        DWORD hashLength = 0;
+        status = BCryptGetProperty(
+            Handle, BCRYPT_HASH_LENGTH, (PBYTE)&hashLength, sizeof(hashLength), &dataLength, 0);
+        if (!BCRYPT_SUCCESS(status))
+        {
+          throw std::runtime_error("failed to call bcrypt function");
+        }
+        HashLength = hashLength;
+      }
+
+      ~AlgorithmProviderInstance() { BCryptCloseAlgorithmProvider(Handle, 0); }
+    };
+
+    static AlgorithmProviderInstance AlgorithmProvider;
+
+    std::string context;
+    context.resize(AlgorithmProvider.ContextSize);
+
+    BCRYPT_HASH_HANDLE hashHandle;
+    NTSTATUS status = BCryptCreateHash(
+        AlgorithmProvider.Handle,
+        &hashHandle,
+        (PUCHAR)context.data(),
+        (ULONG)context.size(),
+        (PUCHAR)key.data(),
+        (ULONG)key.length(),
+        0);
+    if (!BCRYPT_SUCCESS(status))
+    {
+      throw std::runtime_error("failed to call bcrypt function");
+    }
+
+    status = BCryptHashData(hashHandle, (PBYTE)text.data(), (ULONG)text.length(), 0);
+    if (!BCRYPT_SUCCESS(status))
+    {
+      throw std::runtime_error("failed to call bcrypt function");
+    }
+
+    std::string hash;
+    hash.resize(AlgorithmProvider.HashLength);
+    status = BCryptFinishHash(hashHandle, (PUCHAR)hash.data(), (ULONG)hash.length(), 0);
+    if (!BCRYPT_SUCCESS(status))
+    {
+      throw std::runtime_error("failed to call bcrypt function");
+    }
+
+    BCryptDestroyHash(hashHandle);
+
+    return hash;
+  }
+
+  std::string Base64Encode(const std::string& text)
+  {
+    std::string encoded;
+    DWORD encodedLength = DWORD((text.length() + 2) / 3 * 4);
```

Comment:
>DWORD encodedLength = DWORD((text.length() + 2) / 3 * 4); [](start = 4, length = 57)

Please do write some comment here as why the length is calculated this way, for others to be able to maintain this.

Reply:
comments added